src/connector/src/sink/redis.rs

```diff
@@ -12,25 +12,93 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::{HashMap, HashSet};
+
+use anyhow::anyhow;
 use async_trait::async_trait;
+use redis::{Connection, Pipeline};
+use regex::Regex;
 use risingwave_common::array::StreamChunk;
+use risingwave_common::catalog::Schema;
+use serde_derive::Deserialize;
+use serde_with::serde_as;
+
+use super::formatter::SinkFormatterImpl;
+use super::writer::FormattedSink;
+use super::{SinkError, SinkParam, SINK_TYPE_APPEND_ONLY, SINK_TYPE_OPTION, SINK_TYPE_UPSERT};
+use crate::common::RedisCommon;
+use crate::dispatch_sink_formatter_impl;
+use crate::sink::writer::{LogSinkerOf, SinkWriterExt};
+use crate::sink::{DummySinkCommitCoordinator, Result, Sink, SinkWriter, SinkWriterParam};
+
+pub const REDIS_SINK: &str = "redis";
 
-use crate::sink::writer::LogSinkerOf;
-use crate::sink::{
-    DummySinkCommitCoordinator, Result, Sink, SinkError, SinkParam, SinkWriter, SinkWriterParam,
-};
+#[serde_as]
+#[derive(Clone, Debug, Deserialize)]
+pub struct RedisConfig {
+    #[serde(flatten)]
+    pub common: RedisCommon,
 
-#[derive(Clone, Debug)]
-pub struct RedisConfig;
+    pub r#type: String, // accept "append-only" or "upsert"
+}
+
+impl RedisConfig {
+    pub fn from_hashmap(properties: HashMap<String, String>) -> Result<Self> {
+        let config =
+            serde_json::from_value::<RedisConfig>(serde_json::to_value(properties).unwrap())
+                .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        if config.r#type != SINK_TYPE_APPEND_ONLY && config.r#type != SINK_TYPE_UPSERT {
+            return Err(SinkError::Config(anyhow!(
+                "`{}` must be {}, or {}",
+                SINK_TYPE_OPTION,
+                SINK_TYPE_APPEND_ONLY,
+                SINK_TYPE_UPSERT
+            )));
+        }
+        Ok(config)
+    }
+}
 
 #[derive(Debug)]
-pub struct RedisSink;
+pub struct RedisSink {
+    config: RedisConfig,
+    schema: Schema,
+    is_append_only: bool,
+    pk_indices: Vec<usize>,
+}
+
+fn check_string_format(format: &str, set: &HashSet<String>) -> Result<()> {
+    let re = Regex::new(r"\{([^}]*)\}").unwrap();
```

Comment:
will it handle format containing escaped string? such as `'{\"a\": \{{col_a}\}}'`

Reply:
Not yet, escaped characters are treated as strings, and {} can only have column names in it, not other characters.